src/index.ts

```diff
@@ -153,29 +153,67 @@ const getAssetFromKV = async (event: FetchEvent, options?: Partial<Options>): Pr
 
   if (response) {
     let headers = new Headers(response.headers)
-    headers.set('CF-Cache-Status', 'HIT')
     if (shouldSetBrowserCache) {
       headers.set('cache-control', `max-age=${options.cacheControl.browserTTL}`)
     } else {
       // don't assume we want same cache behavior of edge TTL on client
       // so remove the header from the response we'll return
       headers.delete('cache-control')
     }
-    response = new Response(response.body, { headers })
+
+    let shouldRevalidate = false
+    // Four preconditions must be met for a 304 Not Modified:
+    // - the request cannot be a range request
+    // - client sends if-none-match
+    // - resource has etag
+    // - test if-none-match against the pathKey so that we test against KV, rather than against
+    // CF cache, which may modify the etag with a weak validator (e.g. W/"...")
+    shouldRevalidate = [
+      request.headers.has('range') !== true,
+      request.headers.has('if-none-match'),
+      response.headers.has('etag'),
+      request.headers.get('if-none-match') === `${pathKey}`,
+    ].every((val) => val === true)
+
+    if (shouldRevalidate) {
+      // TypeError permitted to satisfy mocks per /issues/96
+      try {
+        // Proactively remove body per /pull/94#discussion_r425455176
+        response.body.cancel()
+      } catch (e) {
+        if (e instanceof TypeError && e.message === 'response.body.cancel is not a function') {
+          console.log('Environment doesnt support readable streams')
+        }
+      }
```

Comment:
```suggestion
      // Proactively remove body per /pull/94#discussion_r425455176
      const cancelResponse = response.body.cancel;
      if (typeof cancelResponse === 'function') {
        cancelResponse();
      } else {
        console.log('Environment doesnt support readable streams');
      }
```

This approach avoid using `try ... catch ...` syntax. With `typeof` is better to know if `response.body.cancel` is a function.

Reply:
Also, there is an idea.

```js
response.headers.set('etag', `${pathKey}`)
```

ETag is only set when:

- No Cache is found from Cloudflare Cache (no `cache.match`).
- `shouldEdgeCache` being `true`.
- Original response has no `etag` header presented.

And the determination of whether return `200 OK, cf-cache-status: hit` or `304 Not Modified, cf-cache-status: revalidate` is based on if `etag` is equal to `pathKey`.

So is it possible to return 304 before trying to get a response from `cache`? I mean, in this way we might be able to bypass `response.body.cancel()` or return the response earlier.

Reply:
Hey @SukkaW
This is certainly good thinking and a strategy I considered before making this PR. Unfortunately the problem is that serving a 304 Not Modified response is more complicated than simply matching the `etag` to `if-none-match`. This is described in great detail here: https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.5.1-13.5.5.

This is sort of a larger topic than can be covered here, but primary reason we need to communicate with cache here is not because of the etag. Rather, it's because the resource stored has the response headers that were served to the client on the request where the etag was granted. Those original response headers need to be included in the 304 Not Modified response. If any of the original response headers are deleted, or any of the "end-to-end" headers like `Transfer-Encoding` are modified, the client will update its version resource because the resource received is different. In other words, when the server sends a 304 Not Modified response, the client expects the resource to not be modified in any way. kv-asset-handler does not store the response headers and the complex tasks of validating etags is beyond the scope of this Worker.

That said, there is nothing stopping you from using kv-asset-handler to return 304s before checking cache :)

Reply:
With respect to your comment on `try...catch`, I don't see any clear performance or logical benefit to changing it to `typeof`. I could be persuaded otherwise I suppose, but I think the more important thing is getting this PR finished. Perhaps bring this back up in the future if you have a good argument for doing so

Reply:
@shagamemnon Agreed. The PR could be merged first.
When will the PR being ready? So I could opened a new PR with merged master branch based.

@victoriabernard92 @harrishancock  Would you guys review the PR as soon as it is ready?